File: 138-copy-list-with-random-pointer/solution.py
from typing import Optional


class Node:
  def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
    self.val = int(x)
    self.next = next
    self.random = random


class Solution:
  def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
    if head is None:
      return None

    node_to_index: dict[Node, int] = {head: 0}
    copy_nodes: list[Node] = [Node(head.val)]
    node = head.next
    index = 1
    while node is not None:
      node_to_index[node] = index
      index += 1

      copy_nodes.append(Node(node.val))
      copy_nodes[-2].next = copy_nodes[-1]
      node = node.next

    node = head
    index = 0
    while node is not None:
      if node.random is not None:
        random_index = node_to_index[node.random]
        copy_nodes[index].random = copy_nodes[random_index]

      node = node.next
      index += 1

    return copy_nodes[0]

File: 138-copy-list-with-random-pointer/solution.cpp
#include <unordered_map>
#include <vector>

class Node
{
public:
  int val;
  Node *next;
  Node *random;

  Node(int _val)
  {
    val = _val;
    next = nullptr;
    random = nullptr;
  }
};

class Solution
{
public:
  Node *copyRandomList(Node *head)
  {
    if (head == nullptr)
    {
      return nullptr;
    }

    std::unordered_map<const Node *, size_t> node_to_index{{head, 0}};
    std::vector<Node *> copy_nodes{new Node(head->val)};
    const auto *node = head->next;
    size_t index = 1;
    while (node != nullptr)
    {
      node_to_index.emplace(node, index);
      ++index;

      copy_nodes.emplace_back(new Node(node->val));
      copy_nodes[copy_nodes.size() - 2]->next = copy_nodes[copy_nodes.size() - 1];
      node = node->next;
    }

    node = head;
    index = 0;
    while (node != nullptr)
    {
      if (const auto iter = node_to_index.find(node->random); iter != node_to_index.end())
      {
        copy_nodes[index]->random = copy_nodes[iter->second];
      }

      node = node->next;
      ++index;
    }

    return copy_nodes[0];
  }
};